text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,396 @@
+
+The timestamp stored in user_key is its inversion result. The purpose is to arrange the newer data (that is, the data with larger timestamp) in front of the older data (that is, data with smaller timestamp). This feature can be utilized in performance optimizing for the data scan process.
+
+#### Data distribution
+
+To facilitate the acquisition and efficient management of multiple versions of a key, the split point of the region will not use the timestamp of the key, so that multiple versions of the key can be guaranteed to be in the same region and the same CF. Considering the size of the region, MaxVerNum will be restrained.
+
+The process of data balancing, scaling, and migration are the same as that of RawKV. In the process of scanning all versions of a key, if the maximum number of versions is exceeded, the redundant versions will be discarded.
+
+#### RPC Interface
+
+With reference to the proto interface, the following interfaces are defined as a new service other than tikv:
+  - Get/BatchGet -  Gets the values for latest version of the key by default 
+  - GetNum/BatchGetNum - Get the values for a specified number of versions of a key. `version_num` must be less than or equal to `MaxVerNum`. 
+  - GetNumWithSV/BatchGetNumWithSV - This interface is like `GetNum/BatchGetNum`, but add requirement about the key with versions ranging from the specified start version to the current version.
+  - GetWithTTL/BatchGetWithTTL - Gets the unexpired values within the TTL.
+  - Scan - Scans the values of all keys of the latest version within a range
+  - ScanNum -  Scans the number of all keys within a range to get the corresponding values of a specified number of versions
+  - ScanNumWithSV - Scans all the keys within a range and gets the corresponding values ranging from the specified start version to the current version 
+  - Put/BatchPut -  Increases the new version value of a key, where version is the current timestamp by default
+  - Delete/BatchDelete - Deletes all versions of a specified key (`key:cur_ts -> delete`)
+  - DeleteRange - Deletes all versions of key in a range using rocksdb DeleteRange
+
+The above fetch operations(get/scan) can use `rocksdb prefix_extractor` to improve the performance of iterator seek. `key: ts-> delete` indicates that all data before the `ts` has been deleted.
+
+### Version recycling
+
+The design of the version recycling will affect cluster performance and the functional designs of incremental backup and CDC. There are two major approaches: 
+  - using multi-version Garbage Collection (GC) similar to TxnKV, and periodically starting the gc worker in the background to delete redundant versions of data;
+  - using compaction filter via rocksdb compaction to delete redundant versions.
```

Comment:
I prefer this way.

Reply:
If compaction filter would have a good gc effect, it is better way.